commit 9beccca0984022a844850e32f0d7dd80d4a225de
Author: Ard Biesheuvel <ardb@kernel.org>
Date:   Thu Oct 27 17:59:07 2022 +0200

    scs: add support for dynamic shadow call stacks
    
    In order to allow arches to use code patching to conditionally emit the
    shadow stack pushes and pops, rather than always taking the performance
    hit even on CPUs that implement alternatives such as stack pointer
    authentication on arm64, add a Kconfig symbol that can be set by the
    arch to omit the SCS codegen itself, without otherwise affecting how
    support code for SCS and compiler options (for register reservation, for
    instance) are emitted.
    
    Also, add a static key and some plumbing to omit the allocation of
    shadow call stack for dynamic SCS configurations if SCS is disabled at
    runtime.
    
    Signed-off-by: Ard Biesheuvel <ardb@kernel.org>
    Reviewed-by: Nick Desaulniers <ndesaulniers@google.com>
    Reviewed-by: Kees Cook <keescook@chromium.org>
    Reviewed-by: Sami Tolvanen <samitolvanen@google.com>
    Tested-by: Sami Tolvanen <samitolvanen@google.com>
    Link: https://lore.kernel.org/r/20221027155908.1940624-3-ardb@kernel.org
    Signed-off-by: Will Deacon <will@kernel.org>

diff --git a/arch/Kconfig b/arch/Kconfig
index 8f138e580d1a..072a1b39e3af 100644
--- a/arch/Kconfig
+++ b/arch/Kconfig
@@ -651,6 +651,13 @@ config SHADOW_CALL_STACK
 	  reading and writing arbitrary memory may be able to locate them
 	  and hijack control flow by modifying the stacks.
 
+config DYNAMIC_SCS
+	bool
+	help
+	  Set by the arch code if it relies on code patching to insert the
+	  shadow call stack push and pop instructions rather than on the
+	  compiler.
+
 config LTO
 	bool
 	help